libvast/vast/transform_step.hpp

```diff
@@ -13,55 +13,33 @@
 #include <arrow/record_batch.h>
 #include <caf/expected.hpp>
 
+#include <queue>
+
 namespace vast {
 
+using arrow_batch
+  = std::tuple<vast::id, vast::type, std::shared_ptr<arrow::RecordBatch>>;
+using batch_queue = std::deque<arrow_batch>;
+using slice_queue = std::deque<table_slice>;
+using batch_vector = std::vector<arrow_batch>;
+
 /// An individual transform step. This is mainly used in the plugin API,
 /// later code deals with a complete `transform`.
 class transform_step {
 public:
   virtual ~transform_step() = default;
 
-  /// Apply the transformation step to the passed table slice,
-  /// if neccessary converting the data to a format the step implementation can
-  /// natively handle.
-  [[nodiscard]] caf::expected<table_slice> apply(table_slice&&) const;
-};
-
-using transform_step_ptr = std::unique_ptr<transform_step>;
+  /// Applies the transformation to a record batch with a corresponding vast
+  /// layout.
+  [[nodiscard]] virtual caf::error
+  add(vast::id offset, type layout, std::shared_ptr<arrow::RecordBatch> batch)
```

Comment:
I still don't understand why the transform step itself needs to know the ID. Conceptually, I think this should be handled in the outer context to reduce the knowledge requirements for the plugin developer.

The only requirement to the transform step should have to be that `finish()` returns at most as many results as the number of calls to `add(slice, layout)` since the last call to `finish()`.

Reply:
The issue is that you'd essentially have to solve the knapsack problem if you want to check this from outside. Let's say a transform step has input ranges `[0,10), [20,40)`. You get back table slices with sizes 8,9,2 and 10. Is this a valid transform or not?

Reply:
There are three scenarios here:

- Import: We can move the ID assignment down into the INDEX stream stage and then we don't need to care about them in the transformations.
- Export: We don't need the IDs at all.
- Partition Transform: We can assign an entirely new ID range to the new partition at once.

Reply:
@lava and I discussed another option, which is to do the re-assigning in the outer context:
- Importer: We can safely assume that there are no gaps in the ID range, so we only need to make sure that the transform engine returned at most as many rows as it got.
- Partition Transformer: Needs to assign a new ID range anyways.
- Exporter: Can safely discard IDs, so no need to do anything.